fileutil/reloader/reloader.go

```diff
@@ -0,0 +1,134 @@
+package reloader
+
+import (
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-phorce/dolly/xlog"
+	"github.com/juju/errors"
+)
+
+var logger = xlog.NewPackageLogger("github.com/go-phorce/dolly/fileutil", "reloader")
+
+// Wrap time.Tick so we can override it in tests.
+var makeTicker = func(interval time.Duration) (func(), <-chan time.Time) {
+	t := time.NewTicker(interval)
+	return t.Stop, t.C
+}
+
+// OnChangedFunc is a called when the file has been modified
+type OnChangedFunc func(filePath string, modifiedAt time.Time)
+
+// Reloader keeps necessary info to provide reloaded certificate
+type Reloader struct {
+	lock           sync.RWMutex
+	loadedAt       time.Time
+	count          uint32
+	filePath       string
+	fileModifiedAt time.Time
+	onChangedFunc  OnChangedFunc
+	inProgress     bool
+	stopChan       chan<- struct{}
+	closed         bool
+}
+
+// NewReloader return an instance of the file re-loader
+func NewReloader(filePath string, checkInterval time.Duration, onChangedFunc OnChangedFunc) (*Reloader, error) {
+	result := &Reloader{
+		filePath:      filePath,
+		onChangedFunc: onChangedFunc,
+		stopChan:      make(chan struct{}),
+	}
+
+	logger.Infof("api=NewReloader, status=started, file=%q", filePath)
+
+	stopChan := make(chan struct{})
+	tickerStop, tickChan := makeTicker(checkInterval)
+	go func() {
+		for {
+			select {
+			case <-stopChan:
+				tickerStop()
+				logger.Infof("api=NewReloader, status=closed, count=%d, file=%q",
+					result.LoadedCount(), filePath)
+				return
+			case <-tickChan:
+				modified := false
+				fi, err := os.Stat(filePath)
+				if err == nil {
+					modified = fi.ModTime().After(result.fileModifiedAt)
+					if modified {
+						result.fileModifiedAt = fi.ModTime()
+						err := result.Reload()
+						if err != nil {
+							logger.Errorf("api=NewReloader, err=[%v]", errors.ErrorStack(err))
+						}
+					}
+				} else {
+					logger.Warningf("api=NewReloader, reason=stat, file=%q, err=[%v]", filePath, err)
+				}
+			}
+		}
+	}()
+	result.stopChan = stopChan
```

Comment:
I wonder if you specifically initiate this at the end? Can't you move it up right after stopChan variable is defined?

Reply:
fixed